// wdio-ui5-service/dist/lib/wdi5-bridge.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
exports._addWdi5Commands = exports.checkForUI5Page = exports.injectUI5 = exports.start = exports.setup = void 0;
var path_1 = require("path");
var promises_1 = require("fs/promises");
var os_1 = require("os");
var semver = __importStar(require("semver"));
var marky_1 = require("marky");
var multiremote_1 = require("webdriverio/build/multiremote");
var wdi5_control_1 = require("./wdi5-control");
var wdi5_fe_1 = require("./wdi5-fe");
var injectTools_1 = require("../../client-side-js/injectTools");
var injectUI5_1 = require("../../client-side-js/injectUI5");
var getSelectorForElement_1 = require("../../client-side-js/getSelectorForElement");
var _checkForUI5Ready_1 = require("../../client-side-js/_checkForUI5Ready");
var getUI5Version_1 = require("../../client-side-js/getUI5Version");
var _navTo_1 = require("../../client-side-js/_navTo");
var allControls_1 = require("../../client-side-js/allControls");
var Logger_1 = require("./Logger");
var Logger = Logger_1.Logger.getInstance();
/** store the status of initialization */
var _isInitialized = false;
/** stores the status of the setup process */
var _setupComplete = false;
/** currently running sap.ui.version */
var _sapUI5Version = null;
/** relay runtime config options from Service */
var _config = null;
function setup(config) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            _config = config;
            if (_setupComplete) {
                // already setup done
                return [2 /*return*/];
            }
            // jump-start the desired log level
            Logger.setLogLevel(config.wdi5.logLevel || "error");
            if (browser instanceof multiremote_1.MultiRemoteDriver) {
                ;
                browser.instances.forEach(function (name) {
                    initBrowser(browser[name]);
                });
                initMultiRemoteBrowser();
            }
            else {
                initBrowser(browser);
            }
            _setupComplete = true;
            return [2 /*return*/];
        });
    });
}
exports.setup = setup;
function start(config) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(config.wdi5.url !== "")) return [3 /*break*/, 2];
                    Logger.info("open url: ".concat(config.wdi5.url));
                    return [4 /*yield*/, browser.url(config.wdi5.url)];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 4];
                case 2:
                    Logger.info("open url with fallback '#' (this is not causing any issues since its is removed for navigation)");
                    return [4 /*yield*/, browser.url("#")];
                case 3:
                    _a.sent();
                    _a.label = 4;
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.start = start;
function initMultiRemoteBrowser() {
    var _this = this;
    ;
    ["asControl", "goTo", "screenshot", "waitForUI5", "getUI5Version", "getSelectorForElement", "allControls"].forEach(function (command) {
        browser.addCommand(command, function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return __awaiter(_this, void 0, void 0, function () {
                var multiRemoteInstance, result;
                var _this = this;
                return __generator(this, function (_a) {
                    multiRemoteInstance = browser;
                    result = [];
                    multiRemoteInstance.instances.forEach(function (name) {
                        result.push(multiRemoteInstance[name][command].apply(_this, args));
                    });
                    return [2 /*return*/, Promise.all(result)];
                });
            });
        });
    });
}
function initBrowser(browserInstance) {
    // init control cache
    if (!browserInstance._controls) {
        Logger.info("creating internal control map");
        browserInstance._controls = [];
    }
    _addWdi5Commands(browserInstance);
    if (!browserInstance.fe) {
        ;
        browserInstance.fe = wdi5_fe_1.WDI5FE;
    }
    _setupComplete = true;
}
function checkUI5Version(ui5Version) {
    if (semver.lt(ui5Version, "1.60.0")) {
        // the record replay api is only available since 1.60
        Logger.error("The ui5 version of your application is to low. Minimum required UI5 version is 1.60");
        throw new Error("The ui5 version of your application is to low. Minimum required UI5 version is 1.60");
    }
}
/**
 * function library to setup the webdriver to UI5 bridge, it runs alle the initial setup
 * make sap/ui/test/RecordReplay accessible via wdio
 * attach the sap/ui/test/RecordReplay object to the application context window object as 'bridge'
 */
function injectUI5(config, browserInstance) {
    return __awaiter(this, void 0, void 0, function () {
        var waitForUI5Timeout, result, version, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    waitForUI5Timeout = config.wdi5.waitForUI5Timeout || 15000;
                    result = true;
                    return [4 /*yield*/, browserInstance.getUI5Version()];
                case 1:
                    version = _b.sent();
                    return [4 /*yield*/, checkUI5Version(version)];
                case 2:
                    _b.sent();
                    return [4 /*yield*/, (0, injectTools_1.clientSide_injectTools)(browserInstance)]; // helpers for wdi5 browser scope
                case 3:
                    _b.sent(); // helpers for wdi5 browser scope
                    _a = result;
                    if (!_a) return [3 /*break*/, 5];
                    return [4 /*yield*/, (0, injectUI5_1.clientSide_injectUI5)(config, waitForUI5Timeout, browserInstance)];
                case 4:
                    _a = (_b.sent());
                    _b.label = 5;
                case 5:
                    result = _a;
                    if (result) {
                        // set when call returns
                        _isInitialized = true;
                        Logger.success("sucessfully initialized wdio-ui5 bridge");
                    }
                    else {
                        Logger.error("bridge was not initialized correctly");
                    }
                    return [2 /*return*/, result];
            }
        });
    });
}
exports.injectUI5 = injectUI5;
function checkForUI5Page() {
    return __awaiter(this, void 0, void 0, function () {
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: 
                // wait till the loading finished and the state is "completed"
                return [4 /*yield*/, browser.waitUntil(function () { return __awaiter(_this, void 0, void 0, function () {
                        var state;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, browser.executeAsync(function (done) {
                                        done(document.readyState);
                                    })];
                                case 1:
                                    state = _a.sent();
                                    return [2 /*return*/, state === "complete"];
                            }
                        });
                    }); })
                    // sap in global window namespace denotes (most likely :) ) that ui5 is present
                ];
                case 1:
                    // wait till the loading finished and the state is "completed"
                    _a.sent();
                    return [4 /*yield*/, browser.executeAsync(function (done) {
                            done(!!window.sap);
                        })];
                case 2: 
                // sap in global window namespace denotes (most likely :) ) that ui5 is present
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.checkForUI5Page = checkForUI5Page;
//******************************************************************************************
// private
/**
 * creates a string valid as object key from a selector
 * @param selector
 * @returns wdio_ui5_key
 */
function _createWdioUI5KeyFromSelector(selector) {
    var orEmpty = function (string) { return string || ""; };
    var _selector = selector.selector;
    var wdi5_ui5_key = "".concat(orEmpty(_selector.id)).concat(orEmpty(_selector.viewName)).concat(orEmpty(_selector.controlType)).concat(orEmpty(JSON.stringify(_selector.bindingPath))).concat(orEmpty(JSON.stringify(_selector.i18NText))).concat(orEmpty(JSON.stringify(_selector.descendant))).concat(orEmpty(JSON.stringify(_selector.labelFor))).concat(orEmpty(JSON.stringify(_selector.properties))).concat(orEmpty(JSON.stringify(_selector.ancestor))).replace(/[^0-9a-zA-Z]+/, "");
    return wdi5_ui5_key;
}
/**
 * does a basic validation of a wdi5ControlSelector
 * @param wdi5Selector: wdi5Selector
 * @returns {boolean} if the given selector is a valid selector
 */
function _verifySelector(wdi5Selector) {
    if (wdi5Selector.hasOwnProperty("selector")) {
        if (wdi5Selector.selector.hasOwnProperty("id") ||
            wdi5Selector.selector.hasOwnProperty("viewName") ||
            wdi5Selector.selector.hasOwnProperty("bindingPath") ||
            wdi5Selector.selector.hasOwnProperty("controlType") ||
            wdi5Selector.selector.hasOwnProperty("i18NText") ||
            wdi5Selector.selector.hasOwnProperty("labelFor") ||
            wdi5Selector.selector.hasOwnProperty("descendant") ||
            wdi5Selector.selector.hasOwnProperty("ancestor") ||
            wdi5Selector.selector.hasOwnProperty("properties") ||
            wdi5Selector.selector.hasOwnProperty("sibling") ||
            wdi5Selector.selector.hasOwnProperty("interactable")) {
            return true;
        }
        Logger.error("Specified selector is not valid. Please use at least one of: 'id, viewName, bindingPath, controlType, i18NText, labelFor, ancestor, properties, descendant, sibling, interactable' -> abort");
        return false;
    }
    Logger.error("Specified selector is not valid -> property 'selector' is missing");
    return false;
}
function _addWdi5Commands(browserInstance) {
    return __awaiter(this, void 0, void 0, function () {
        var _this = this;
        return __generator(this, function (_a) {
            browserInstance.addCommand("_asControl", function (wdi5Selector) { return __awaiter(_this, void 0, void 0, function () {
                var internalKey, wdi5Control, e;
                var _a;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            if (!_verifySelector(wdi5Selector)) {
                                return [2 /*return*/, "ERROR: Specified selector is not valid -> abort"];
                            }
                            internalKey = wdi5Selector.wdio_ui5_key || _createWdioUI5KeyFromSelector(wdi5Selector);
                            if (!(!((_a = browserInstance._controls) === null || _a === void 0 ? void 0 : _a[internalKey]) || wdi5Selector.forceSelect) /* always retrieve control */) return [3 /*break*/, 2]; /* always retrieve control */
                            Logger.info("creating internal control with id ".concat(internalKey));
                            wdi5Selector.wdio_ui5_key = internalKey;
                            (0, marky_1.mark)("retrieveSingleControl");
                            return [4 /*yield*/, new wdi5_control_1.WDI5Control({ browserInstance: browserInstance }).init(wdi5Selector, wdi5Selector.forceSelect)];
                        case 1:
                            wdi5Control = _b.sent();
                            e = (0, marky_1.stop)("retrieveSingleControl");
                            Logger.info("_asControl() needed ".concat(e.duration, " for ").concat(internalKey));
                            browserInstance._controls[internalKey] = wdi5Control;
                            return [3 /*break*/, 3];
                        case 2:
                            Logger.info("reusing internal control with id ".concat(internalKey));
                            _b.label = 3;
                        case 3: return [2 /*return*/, browserInstance._controls[internalKey]];
                    }
                });
            }); });
            // no fluent API -> no private method
            browserInstance.addCommand("allControls", function (wdi5Selector) { return __awaiter(_this, void 0, void 0, function () {
                var internalKey, _a, _b;
                var _c;
                return __generator(this, function (_d) {
                    switch (_d.label) {
                        case 0:
                            if (!_verifySelector(wdi5Selector)) {
                                return [2 /*return*/, "ERROR: Specified selector is not valid -> abort"];
                            }
                            internalKey = wdi5Selector.wdio_ui5_key || _createWdioUI5KeyFromSelector(wdi5Selector);
                            if (!(!((_c = browserInstance._controls) === null || _c === void 0 ? void 0 : _c[internalKey]) || wdi5Selector.forceSelect) /* always retrieve control */) return [3 /*break*/, 2]; /* always retrieve control */
                            wdi5Selector.wdio_ui5_key = internalKey;
                            Logger.info("creating internal controls with id ".concat(internalKey));
                            _a = browserInstance._controls;
                            _b = internalKey;
                            return [4 /*yield*/, _allControls(wdi5Selector, browserInstance)];
                        case 1:
                            _a[_b] = _d.sent();
                            return [2 /*return*/, browserInstance._controls[internalKey]];
                        case 2:
                            Logger.info("reusing internal control with id ".concat(internalKey));
                            _d.label = 3;
                        case 3: return [2 /*return*/, browserInstance._controls[internalKey]];
                    }
                });
            }); });
            /**
             * Find the best control selector for a DOM element. A selector uniquely represents a single element.
             * The 'best' selector is the one with which it is most likely to uniquely identify a control with the least possible inspection of the control tree.
             * @param {object} oOptions
             * @param {object} oOptions.domElement - DOM Element to search for
             * @param {object} oOptions.settings - ui5 settings object
             * @param {boolean} oOptions.settings.preferViewId
             */
            browserInstance.addCommand("getSelectorForElement", function (oOptions) { return __awaiter(_this, void 0, void 0, function () {
                var result;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, (0, getSelectorForElement_1.clientSide_getSelectorForElement)(oOptions, browserInstance)];
                        case 1:
                            result = (_a.sent());
                            if (result.status === 1) {
                                console.error("ERROR: getSelectorForElement() failed because of: " + result.message);
                                return [2 /*return*/, result.message];
                            }
                            else if (result.status === 0) {
                                console.log("SUCCESS: getSelectorForElement() returned:  ".concat(JSON.stringify(result.result)));
                                return [2 /*return*/, result.result];
                            }
                            return [2 /*return*/];
                    }
                });
            }); });
            browserInstance.addCommand("getUI5Version", function () { return __awaiter(_this, void 0, void 0, function () {
                var resultVersion;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!!_sapUI5Version) return [3 /*break*/, 2];
                            return [4 /*yield*/, (0, getUI5Version_1.clientSide_getUI5Version)(browserInstance)];
                        case 1:
                            resultVersion = _a.sent();
                            _sapUI5Version = resultVersion;
                            _a.label = 2;
                        case 2: return [2 /*return*/, _sapUI5Version];
                    }
                });
            }); });
            /**
             * uses the UI5 native waitForUI5 function to wait for all promises to be settled
             */
            browserInstance.addCommand("waitForUI5", function () { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, _waitForUI5(browserInstance)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            }); });
            /**
             * wait for ui5 and take a screenshot
             */
            browserInstance.addCommand("screenshot", function (fileAppendix) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, _waitForUI5(browserInstance)];
                        case 1:
                            _a.sent();
                            return [4 /*yield*/, _writeScreenshot(fileAppendix)];
                        case 2:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            }); });
            browserInstance.addCommand("goTo", function (oOptions) { return __awaiter(_this, void 0, void 0, function () {
                var sHash, oRoute, url, _a, currentUrl, alreadyNavByHash, navToRoot;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            if (typeof oOptions === "string") {
                                sHash = oOptions;
                            }
                            else {
                                sHash = oOptions.sHash;
                            }
                            oRoute = oOptions.oRoute;
                            if (!(sHash && sHash.length > 0)) return [3 /*break*/, 13];
                            _a = browserInstance.config.wdi5["url"];
                            if (_a) return [3 /*break*/, 2];
                            return [4 /*yield*/, browserInstance.getUrl()];
                        case 1:
                            _a = (_b.sent());
                            _b.label = 2;
                        case 2:
                            url = _a;
                            if (!(url && url.length > 0 && url !== "#")) return [3 /*break*/, 8];
                            return [4 /*yield*/, browserInstance.getUrl()];
                        case 3:
                            currentUrl = _b.sent();
                            alreadyNavByHash = currentUrl.includes("#");
                            navToRoot = url.startsWith("/");
                            if (!(alreadyNavByHash && !navToRoot)) return [3 /*break*/, 5];
                            return [4 /*yield*/, browserInstance.url("".concat(currentUrl.split("#")[0]).concat(sHash))];
                        case 4:
                            _b.sent();
                            return [3 /*break*/, 7];
                        case 5: return [4 /*yield*/, browserInstance.url("".concat(url).concat(sHash))];
                        case 6:
                            _b.sent();
                            _b.label = 7;
                        case 7: return [3 /*break*/, 12];
                        case 8:
                            if (!(url && url.length > 0 && url === "#")) return [3 /*break*/, 10];
                            // route without the double hash
                            return [4 /*yield*/, browserInstance.url("".concat(sHash))];
                        case 9:
                            // route without the double hash
                            _b.sent();
                            return [3 /*break*/, 12];
                        case 10: 
                        // just a fallback
                        return [4 /*yield*/, browserInstance.url("".concat(sHash))];
                        case 11:
                            // just a fallback
                            _b.sent();
                            _b.label = 12;
                        case 12: return [3 /*break*/, 16];
                        case 13:
                            if (!(oRoute && oRoute.sName)) return [3 /*break*/, 15];
                            // navigate using the ui5 router
                            // sComponentId, sName, oParameters, oComponentTargetInfo, bReplace
                            return [4 /*yield*/, _navTo(oRoute.sComponentId, oRoute.sName, oRoute.oParameters, oRoute.oComponentTargetInfo, oRoute.bReplace, browserInstance)];
                        case 14:
                            // navigate using the ui5 router
                            // sComponentId, sName, oParameters, oComponentTargetInfo, bReplace
                            _b.sent();
                            return [3 /*break*/, 16];
                        case 15:
                            Logger.error("ERROR: navigating to another page");
                            _b.label = 16;
                        case 16: return [2 /*return*/];
                    }
                });
            }); });
            // inspired by and after staring a long time hard at:
            // https://stackoverflow.com/questions/51635378/keep-object-chainable-using-async-methods
            // https://github.com/Shigma/prochain
            // https://github.com/l8js/l8/blob/main/src/core/liquify.js
            // channel the async function browser._asControl (init'ed via browser.addCommand above) through a Proxy
            // in order to chain calls of any subsequent UI5 api methods on the retrieved UI5 control:
            // await browser.asControl(selector).methodOfUI5control().anotherMethodOfUI5control()
            // the way this works is twofold:
            // 1. (almost) all UI5 $control's API methods are reinjected from the browser-scope
            //    into the Node.js scope via async WDI5._executeControlMethod(), which in term actually calls
            //    the reinjected API method within the browser scope
            // 2. the execution of each UI5 $control's API method (via async WDI5._executeControlMethod() => Promise) is then chained
            //    via the below "then"-ing of the (async WDI5._executeControlMethod() => Promise)-Promises with the help of
            //    the a Proxy and a recursive `handler` function
            if (!browserInstance.asControl) {
                browserInstance.asControl = function (ui5ControlSelector) {
                    var _a;
                    var asyncMethods = ["then", "catch", "finally"];
                    var functionQueue = [];
                    // we need to do the same operation as in the 'init' of 'wdi5-control.ts'
                    var logging = (_a = ui5ControlSelector === null || ui5ControlSelector === void 0 ? void 0 : ui5ControlSelector.logging) !== null && _a !== void 0 ? _a : true;
                    function makeFluent(target) {
                        var promise = Promise.resolve(target);
                        var handler = {
                            get: function (_, prop) {
                                functionQueue.push(prop);
                                return asyncMethods.includes(prop)
                                    ? function () {
                                        var boundArgs = [];
                                        for (var _i = 0; _i < arguments.length; _i++) {
                                            boundArgs[_i] = arguments[_i];
                                        }
                                        return makeFluent(promise[prop].apply(promise, __spreadArray([], __read(boundArgs), false)));
                                    }
                                    : makeFluent(promise.then(function (object) {
                                        // when object is undefined the previous function call failed
                                        try {
                                            return object[prop];
                                        }
                                        catch (error) {
                                            // different node versions return a different `error.message` so we use our own message
                                            if (logging) {
                                                Logger.error("Cannot read property '".concat(prop, "' in the execution queue!"));
                                            }
                                        }
                                    }));
                            },
                            apply: function (_, thisArg, boundArgs) {
                                return makeFluent(
                                // When "targetFunction" is empty we can assume that there are errors in the execution queue
                                promise.then(function (targetFunction) {
                                    if (targetFunction) {
                                        return Reflect.apply(targetFunction, thisArg, boundArgs);
                                    }
                                    else {
                                        // a functionQueue without a 'then' can be ignored
                                        // as the original error was already logged
                                        if (functionQueue.includes("then") && logging) {
                                            functionQueue.splice(functionQueue.indexOf("then"));
                                            Logger.error("One of the calls in the queue \"".concat(functionQueue.join("()."), "()\" previously failed!"));
                                        }
                                    }
                                }));
                            }
                        };
                        // eslint-disable-next-line @typescript-eslint/no-empty-function
                        return new Proxy(function () { }, handler);
                    }
                    // @ts-ignore
                    return makeFluent(browserInstance._asControl(ui5ControlSelector));
                };
            }
            return [2 /*return*/];
        });
    });
}
exports._addWdi5Commands = _addWdi5Commands;
/**
 * retrieve a DOM element via UI5 locator
 * @param {sap.ui.test.RecordReplay.ControlSelector} controlSelector
 * @return {[WebdriverIO.Element | String, [aProtoFunctions]]} UI5 control or error message, array of function names of this control
 */
function _allControls(controlSelector, browserInstance) {
    var e_1, _a;
    var _b, _c;
    if (controlSelector === void 0) { controlSelector = this._controlSelector; }
    if (browserInstance === void 0) { browserInstance = browser; }
    return __awaiter(this, void 0, void 0, function () {
        var response, retrievedElements, resultWDi5Elements, retrievedElements_1, retrievedElements_1_1, cControl, oOptions, e_1_1;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    // check whether we have a "by id regex" locator request
                    if (controlSelector.selector.id && typeof controlSelector.selector.id === "object") {
                        // make it a string for serializing into browser-scope and
                        // further processing there
                        controlSelector.selector.id = controlSelector.selector.id.toString();
                    }
                    if (typeof ((_b = controlSelector.selector.properties) === null || _b === void 0 ? void 0 : _b.text) === "object" &&
                        ((_c = controlSelector.selector.properties) === null || _c === void 0 ? void 0 : _c.text) instanceof RegExp) {
                        // make it a string for serializing into browser-scope and
                        // further processing there
                        controlSelector.selector.properties.text = controlSelector.selector.properties.text.toString();
                    }
                    return [4 /*yield*/, (0, allControls_1.clientSide_allControls)(controlSelector, browserInstance)];
                case 1:
                    response = _d.sent();
                    _writeObjectResultLog(response, "allControls()");
                    if (!(response.status === 0)) return [3 /*break*/, 14];
                    retrievedElements = response.result;
                    resultWDi5Elements = [];
                    _d.label = 2;
                case 2:
                    _d.trys.push([2, 7, 8, 13]);
                    retrievedElements_1 = __asyncValues(retrievedElements);
                    _d.label = 3;
                case 3: return [4 /*yield*/, retrievedElements_1.next()];
                case 4:
                    if (!(retrievedElements_1_1 = _d.sent(), !retrievedElements_1_1.done)) return [3 /*break*/, 6];
                    cControl = retrievedElements_1_1.value;
                    oOptions = {
                        controlSelector: controlSelector,
                        wdio_ui5_key: controlSelector.wdio_ui5_key,
                        forceSelect: controlSelector.forceSelect,
                        generatedUI5Methods: cControl.aProtoFunctions,
                        webdriverRepresentation: null,
                        webElement: cControl.domElement,
                        domId: cControl.id,
                        browserInstance: browserInstance
                    };
                    // FIXME: multi remote support by providing browserInstance in constructor
                    resultWDi5Elements.push(new wdi5_control_1.WDI5Control(oOptions));
                    _d.label = 5;
                case 5: return [3 /*break*/, 3];
                case 6: return [3 /*break*/, 13];
                case 7:
                    e_1_1 = _d.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 13];
                case 8:
                    _d.trys.push([8, , 11, 12]);
                    if (!(retrievedElements_1_1 && !retrievedElements_1_1.done && (_a = retrievedElements_1["return"]))) return [3 /*break*/, 10];
                    return [4 /*yield*/, _a.call(retrievedElements_1)];
                case 9:
                    _d.sent();
                    _d.label = 10;
                case 10: return [3 /*break*/, 12];
                case 11:
                    if (e_1) throw e_1.error;
                    return [7 /*endfinally*/];
                case 12: return [7 /*endfinally*/];
                case 13: return [2 /*return*/, resultWDi5Elements];
                case 14: return [2 /*return*/, "[WDI5] Error: fetch multiple elements failed: " + response.message];
            }
        });
    });
}
/**
 * can be called to make sure before you access any eg. DOM Node the ui5 framework is done loading
 * @returns {Boolean} if the UI5 page is fully loaded and ready to interact.
 */
function _waitForUI5(browserInstance) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!_isInitialized) return [3 /*break*/, 2];
                    return [4 /*yield*/, _checkForUI5Ready(browserInstance)];
                case 1: 
                // injectUI5 was already called and was successful attached
                return [2 /*return*/, _a.sent()];
                case 2: return [4 /*yield*/, injectUI5(_config, browserInstance)];
                case 3:
                    if (!_a.sent()) return [3 /*break*/, 5];
                    return [4 /*yield*/, _checkForUI5Ready(browserInstance)];
                case 4: return [2 /*return*/, _a.sent()];
                case 5: return [2 /*return*/, false];
            }
        });
    });
}
/**
 * check for UI5 via the RecordReplay.waitForUI5 method
 */
function _checkForUI5Ready(browserInstance) {
    return __awaiter(this, void 0, void 0, function () {
        var ready;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    ready = false;
                    if (!_isInitialized) return [3 /*break*/, 2];
                    return [4 /*yield*/, (0, _checkForUI5Ready_1.clientSide__checkForUI5Ready)(browserInstance)];
                case 1: 
                // can only be executed when RecordReplay is attached
                return [2 /*return*/, _a.sent()];
                case 2: return [2 /*return*/, ready];
            }
        });
    });
}
/**
 * @param fileAppendix
 */
function _writeScreenshot(fileAppendix) {
    if (fileAppendix === void 0) { fileAppendix = "-screenshot"; }
    return __awaiter(this, void 0, void 0, function () {
        var screenshot, seed, _path, path, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    // if config param screenshotsDisabled is set to true -> no screenshots will be taken
                    if (_config.wdi5["screenshotsDisabled"]) {
                        Logger.warn("screenshot skipped due to config parameter");
                        return [2 /*return*/];
                    }
                    return [4 /*yield*/, browser.takeScreenshot()];
                case 1:
                    screenshot = _a.sent();
                    seed = _getDateString();
                    _path = _config.wdi5.screenshotPath || (0, os_1.tmpdir)();
                    path = (0, path_1.resolve)(_path, "".concat(seed, "-").concat(fileAppendix, ".png"));
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, 4, , 5]);
                    return [4 /*yield*/, (0, promises_1.writeFile)(path, screenshot, "base64")];
                case 3:
                    _a.sent();
                    Logger.success("screenshot at ".concat(path, " created"));
                    return [3 /*break*/, 5];
                case 4:
                    error_1 = _a.sent();
                    Logger.error("error while saving screenshot: ".concat(error_1));
                    return [3 /*break*/, 5];
                case 5: return [2 /*return*/];
            }
        });
    });
}
/**
 * generates date string with format M-d-hh-mm-ss
 * @returns {String}
 */
function _getDateString() {
    var x = new Date();
    return "".concat(x.getMonth() + 1, "-").concat(x.getDate(), "-").concat(x.getHours(), "-").concat(x.getMinutes(), "-").concat(x.getSeconds());
}
/**
 * navigates to a UI5 route using the Component router
 * @param {String} sComponentId
 * @param {String} sName
 * @param {Object} oParameters
 * @param {Object} oComponentTargetInfo
 * @param {Boolean} bReplace
 */
function _navTo(sComponentId, sName, oParameters, oComponentTargetInfo, bReplace, browserInstance) {
    return __awaiter(this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, (0, _navTo_1.clientSide__navTo)(sComponentId, sName, oParameters, oComponentTargetInfo, bReplace, browserInstance)];
                case 1:
                    result = (_a.sent());
                    if (result.status === 1) {
                        Logger.error("ERROR: navigation using UI5 router failed because of: " + result.message);
                        return [2 /*return*/, result.result];
                    }
                    else if (result.status === 0) {
                        Logger.info("SUCCESS: navigation using UI5 router to hash: ".concat(JSON.stringify(result.status)));
                        return [2 /*return*/, result.result];
                    }
                    return [2 /*return*/];
            }
        });
    });
}
/**
 * create log based on the status of result.status
 * @param {Array} result
 * @param {*} functionName
 */
function _writeObjectResultLog(response, functionName) {
    if (response.status > 0) {
        Logger.error("call of ".concat(functionName, " failed because of: ").concat(response.message));
    }
    else if (response.status === 0) {
        Logger.success("call of function ".concat(functionName, " returned: ").concat(JSON.stringify(response.id ? response.id : response.result)));
    }
    else {
        Logger.warn("Unknown status: ".concat(functionName, " returned: ").concat(JSON.stringify(response.message)));
    }
}
